clients/ts/signalr/src/HttpConnection.ts

```diff
@@ -123,16 +113,50 @@ export class HttpConnection implements IConnection {
         }
     }
 
-    private createTransport(requestedTransport: TransportType | ITransport, availableTransports: IAvailableTransport[], requestedTransferFormat: TransferFormat): ITransport {
+    private async setNegotiationResponse(headers: any): Promise<INegotiateResponse> {
+        const response =  await this.httpClient.post(this.resolveNegotiateUrl(this.baseUrl), {
+            content: "",
+            headers,
+        });
+        return JSON.parse(response.content as string);
+    }
+
+    private updateConnectionId(negotiateResponse: INegotiateResponse) {
+        this.connectionId = negotiateResponse.connectionId;
+        this.url = this.baseUrl + (this.baseUrl.indexOf("?") === -1 ? "?" : "&") + `id=${this.connectionId}`;
+    }
+
+    private async createTransport(requestedTransport: TransportType | ITransport, negotiateResponse: INegotiateResponse, requestedTransferFormat: TransferFormat, headers: any): Promise<void> {
+        this.updateConnectionId(negotiateResponse);
         if (this.isITransport(requestedTransport)) {
             this.logger.log(LogLevel.Trace, "Connection was provided an instance of ITransport, using that directly.");
-            return requestedTransport;
+            this.transport = requestedTransport;
+            await this.transport.connect(this.url, requestedTransferFormat, this);
+
+            // only change the state if we were connecting to not overwrite
+            // the state if the connection is already marked as Disconnected
+            this.changeState(ConnectionState.Connecting, ConnectionState.Connected);
+            return Promise.resolve();
```

Comment:
`async` function means just use `return` here. TypeScript manages the outer `Promise`